# Lending Club Loan Data Understanding

## Objective

This notebook performs initial data understanding for the Lending Club loan dataset.

The goal is to inspect the dataset structure, identify important columns, understand missing values, review data types, and examine the target variable before building borrower risk and loan default models.

This notebook covers:
- Loading the dataset
- Reviewing column structure
- Checking duplicates
- Understanding the target variable
- Data type summary
- Missing value analysis
- Numeric and categorical summaries
- Initial column overview

## Import Required Libraries

Pandas and NumPy are used for data loading, inspection, and basic data analysis.

In [13]:
import pandas as pd
import numpy as np

In [14]:
pd.set_option('display.max_columns',100)

## Load Lending Club Dataset

The Lending Club loan dataset is loaded from the local `data` folder.

The dataset contains loan-level borrower, credit, repayment, and recovery information.

In [15]:
df=pd.read_csv("../data/loan.csv", low_memory=True)
df.head()

/var/folders/99/nqyz4s796bx09s6p71mvxd5h0000gn/T/ipykernel_41842/251095567.py:1: DtypeWarning: Columns (19,47,55,112,123,124,125,128,129,130,133,139,140,141) have mixed types. Specify dtype option on import or set low_memory=False.
  df=pd.read_csv("../data/loan.csv", low_memory=True)


,id,member_id,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,emp_title,emp_length,home_ownership,annual_inc,verification_status,issue_d,loan_status,pymnt_plan,url,desc,purpose,title,zip_code,addr_state,dti,delinq_2yrs,earliest_cr_line,inq_last_6mths,mths_since_last_delinq,mths_since_last_record,open_acc,pub_rec,revol_bal,revol_util,total_acc,initial_list_status,out_prncp,out_prncp_inv,total_pymnt,total_pymnt_inv,total_rec_prncp,total_rec_int,total_rec_late_fee,recoveries,collection_recovery_fee,last_pymnt_d,last_pymnt_amnt,next_pymnt_d,last_credit_pull_d,collections_12_mths_ex_med,...,num_op_rev_tl,num_rev_accts,num_rev_tl_bal_gt_0,num_sats,num_tl_120dpd_2m,num_tl_30dpd,num_tl_90g_dpd_24m,num_tl_op_past_12m,pct_tl_nvr_dlq,percent_bc_gt_75,pub_rec_bankruptcies,tax_liens,tot_hi_cred_lim,total_bal_ex_mort,total_bc_limit,total_il_high_credit_limit,revol_bal_joint,sec_app_earliest_cr_line,sec_app_inq_last_6mths,sec_app_mort_acc,sec_app_open_acc,sec_app_revol_util,sec_app_open_act_il,sec_app_num_rev_accts,sec_app_chargeoff_within_12_mths,sec_app_collections_12_mths_ex_med,sec_app_mths_since_last_major_derog,hardship_flag,hardship_type,hardship_reason,hardship_status,deferral_term,hardship_amount,hardship_start_date,hardship_end_date,payment_plan_start_date,hardship_length,hardship_dpd,hardship_loan_status,orig_projected_additional_accrued_interest,hardship_payoff_balance_amount,hardship_last_payment_amount,disbursement_method,debt_settlement_flag,debt_settlement_flag_date,settlement_status,settlement_date,settlement_amount,settlement_percentage,settlement_term
0,NaN,NaN,2500,2500,2500.0,36 months,13.56,84.92,C,C1,Chef,10+ years,RENT,55000.0,Not Verified,Dec-2018,Current,n,NaN,NaN,debt_consolidation,Debt consolidation,109xx,NY,18.24,0.0,Apr-2001,1.0,NaN,45.0,9.0,1.0,4341,10.3,34.0,w,2386.02,2386.02,167.02,167.02,113.98,53.04,0.0,0.0,0.0,Feb-2019,84.92,Mar-2019,Feb-2019,0.0,...,7.0,18.0,5.0,9.0,0.0,0.0,0.0,3.0,100.0,0.0,1.0,0.0,60124.0,16901.0,36500.0,18124.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,NaN,30000,30000,30000.0,60 months,18.94,777.23,D,D2,Postmaster,10+ years,MORTGAGE,90000.0,Source Verified,Dec-2018,Current,n,NaN,NaN,debt_consolidation,Debt consolidation,713xx,LA,26.52,0.0,Jun-1987,0.0,71.0,75.0,13.0,1.0,12315,24.2,44.0,w,29387.75,29387.75,1507.11,1507.11,612.25,894.86,0.0,0.0,0.0,Feb-2019,777.23,Mar-2019,Feb-2019,0.0,...,8.0,14.0,4.0,13.0,0.0,0.0,0.0,6.0,95.0,0.0,1.0,0.0,372872.0,99468.0,15000.0,94072.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,5000,5000,5000.0,36 months,17.97,180.69,D,D1,Administrative,6 years,MORTGAGE,59280.0,Source Verified,Dec-2018,Current,n,NaN,NaN,debt_consolidation,Debt consolidation,490xx,MI,10.51,0.0,Apr-2011,0.0,NaN,NaN,8.0,0.0,4599,19.1,13.0,w,4787.21,4787.21,353.89,353.89,212.79,141.10,0.0,0.0,0.0,Feb-2019,180.69,Mar-2019,Feb-2019,0.0,...,6.0,7.0,3.0,8.0,0.0,0.0,0.0,0.0,100.0,0.0,0.0,0.0,136927.0,11749.0,13800.0,10000.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
3,NaN,NaN,4000,4000,4000.0,36 months,18.94,146.51,D,D2,IT Supervisor,10+ years,MORTGAGE,92000.0,Source Verified,Dec-2018,Current,n,NaN,NaN,debt_consolidation,Debt consolidation,985xx,WA,16.74,0.0,Feb-2006,0.0,NaN,NaN,10.0,0.0,5468,78.1,13.0,w,3831.93,3831.93,286.71,286.71,168.07,118.64,0.0,0.0,0.0,Feb-2019,146.51,Mar-2019,Feb-2019,0.0,...,2.0,3.0,2.0,10.0,0.0,0.0,0.0,3.0,100.0,100.0,0.0,0.0,385183.0,36151.0,5000.0,44984.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,N,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
4,NaN,NaN,30000,30000,30000.0,60 months,16.14,731.78,C,C4,Mechanic,10+ years,MORTGAGE,57250.0,Not Verified,Dec-2018,Current,n,NaN,NaN,debt_consolidation,Debt consolidati

## Review Available Columns

The dataset contains a large number of borrower, loan, credit history, payment, hardship, and settlement-related columns.

Reviewing column names helps understand what information is available for future feature engineering and modeling.

In [16]:
df.columns.to_list()

['id',
 'member_id',
 'loan_amnt',
 'funded_amnt',
 'funded_amnt_inv',
 'term',
 'int_rate',
 'installment',
 'grade',
 'sub_grade',
 'emp_title',
 'emp_length',
 'home_ownership',
 'annual_inc',
 'verification_status',
 'issue_d',
 'loan_status',
 'pymnt_plan',
 'url',
 'desc',
 'purpose',
 'title',
 'zip_code',
 'addr_state',
 'dti',
 'delinq_2yrs',
 'earliest_cr_line',
 'inq_last_6mths',
 'mths_since_last_delinq',
 'mths_since_last_record',
 'open_acc',
 'pub_rec',
 'revol_bal',
 'revol_util',
 'total_acc',
 'initial_list_status',
 'out_prncp',
 'out_prncp_inv',
 'total_pymnt',
 'total_pymnt_inv',
 'total_rec_prncp',
 'total_rec_int',
 'total_rec_late_fee',
 'recoveries',
 'collection_recovery_fee',
 'last_pymnt_d',
 'last_pymnt_amnt',
 'next_pymnt_d',
 'last_credit_pull_d',
 'collections_12_mths_ex_med',
 'mths_since_last_major_derog',
 'policy_code',
 'application_type',
 'annual_inc_joint',
 'dti_joint',
 'verification_status_joint',
 'acc_now_delinq',
 'tot_coll_amt',
 'tot_cur_

## Dataset Structure

The `info()` function is used to understand:

- number of rows
- number of columns
- column data types
- memory usage

This is important because the dataset is large and contains mixed data types.

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2260668 entries, 0 to 2260667
Columns: 145 entries, id to settlement_term
dtypes: float64(105), int64(4), object(36)
memory usage: 2.4+ GB


## Duplicate Record Check

Duplicate rows are checked to ensure that the dataset does not contain repeated loan records.

Removing duplicates is important before model development because duplicate records can bias analysis and model training.

In [19]:
#Checking for duplicated
duplicated_count=df.duplicated().sum()
print(f'Duplicated Count : {duplicated_count}')

Duplicated Count : 0


## Target Variable Distribution

The `loan_status` column is inspected because it will be used to define repayment outcome or default behavior.

Understanding the distribution of loan statuses helps decide how to convert multi-class loan outcomes into a binary target for risk modeling.

In [21]:
#Target Variable Distribution
df['loan_status'].value_counts(dropna=False)

loan_status
Fully Paid                                             1041952
Current                                                 919695
Charged Off                                             261655
Late (31-120 days)                                       21897
In Grace Period                                           8952
Late (16-30 days)                                         3737
Does not meet the credit policy. Status:Fully Paid        1988
Does not meet the credit policy. Status:Charged Off        761
Default                                                     31
Name: count, dtype: int64

## Data Type Summary

The dataset contains numerical and categorical variables.

Understanding the number of each data type helps plan preprocessing steps such as encoding categorical variables and scaling numerical variables.

In [23]:
#Data Type Summary
df.dtypes.value_counts()

float64    105
object      36
int64        4
Name: count, dtype: int64

## Missing Value Analysis

Missing values are calculated for each column.

Columns with very high missing percentages may need to be removed because they provide little usable information for modeling.

This step helps identify:
- completely empty columns
- sparse hardship or settlement columns
- columns requiring imputation

In [29]:
missing_summary= df.isnull().sum().to_frame('missing_count')\
    .assign(missing_percent=lambda x: (x['missing_count']/len(df))*100)\
    .sort_values('missing_percent',ascending=False)

missing_summary.head(30)

,missing_count,missing_percent
id,2260668,100.000000
url,2260668,100.000000
member_id,2260668,100.000000
orig_projected_additional_accrued_interest,2252242,99.627278
hardship_length,2250055,99.530537
hardship_reason,2250055,99.530537
hardship_status,2250055,99.530537
deferral_term,2250055,99.530537
hardship_amount,2250055,99.530537
hardship_start_date,2250055,99.530537


## Numerical Feature Summary

Summary statistics are generated for numerical columns.

This helps identify:
- minimum and maximum values
- average values
- spread of the data
- potential outliers
- unusual values such as zero income or extreme debt-to-income ratio

In [30]:
# Basic Numeric Summary
df.describe().T.head(30)

,count,mean,std,min,25%,50%,75%,max
id,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
member_id,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
loan_amnt,2260668.0,15046.931228,9190.245488,5.000000e+02,8000.00,12900.000000,20000.000000,4.000000e+04
funded_amnt,2260668.0,15041.664057,9188.413022,5.000000e+02,8000.00,12875.000000,20000.000000,4.000000e+04
funded_amnt_inv,2260668.0,15023.437624,9192.331807,0.000000e+00,8000.00,12800.000000,20000.000000,4.000000e+04
int_rate,2260668.0,13.092913,4.832114,5.310000e+00,9.49,12.620000,15.990000,3.099000e+01
installment,2260668.0,445.807646,267.173725,4.930000e+00,251.65,377.990000,593.320000,1.719830e+03
annual_inc,2260664.0,77992.428687,112696.199574,0.000000e+00,46000.00,65000.000000,93000.000000,1.100000e+08
url,0.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
dti,2258957.0,18.824196,14.183329,-1.000000e+00,11.89,17.840000,24.490000,9.990000e+02


## Categorical and Numerical Column Split

Columns are separated into categorical and numerical groups.

This separation is useful because each type requires different preprocessing:

- categorical columns may need encoding
- numerical columns may need scaling, imputation, or outlier handling

In [33]:
# Basic Object/categorical summary
cat_cols = df.select_dtypes(include='object').columns.tolist()
num_cols = df.select_dtypes(exclude='object').columns.tolist()

print(f'Categorical Columns:', len(cat_cols))
print(f'Numeric Columns:', len(num_cols))

Categorical Columns: 36
Numeric Columns: 109


## Initial Important Column Selection

A preliminary list of important columns is created based on domain relevance.

These columns include borrower income, loan amount, interest rate, grade, loan status, credit history, revolving balance, recoveries, and principal repayment information.

This list helps guide future borrower profiling and credit risk modeling.

In [34]:
important_cols = ["loan_amnt", "funded_amnt", "term", "int_rate", "installment","grade", "sub_grade", "emp_length", "home_ownership", 
                  "annual_inc", "verification_status", "loan_status","purpose", "dti", "fico_range_low", "fico_range_high","open_acc",
                  "pub_rec", "revol_bal", "revol_util","total_acc", "recoveries", "total_rec_prncp","out_prncp"]


## Column-Level Data Quality Overview

A column overview table is created to summarize:

- column name
- data type
- missing count
- missing percentage
- number of unique values

This provides a high-level data quality report and helps decide which columns to keep, drop, or transform in later notebooks.

In [36]:
columns_overview = pd.DataFrame({
    'column_name': df.columns,
    'dtype': df.dtypes.astype(str).values,
    'missing_count': df.isnull().sum().values,
    'missing_percent': (df.isnull().sum().values/len(df))*100,
    'unique_values': df.nunique(dropna=False).values
    
})

columns_overview=columns_overview.sort_values('missing_percent',ascending=False)
columns_overview.head(30)

,column_name,dtype,missing_count,missing_percent,unique_values
0,id,float64,2260668,100.000000,1
18,url,float64,2260668,100.000000,1
1,member_id,float64,2260668,100.000000,1
134,orig_projected_additional_accrued_interest,float64,2252242,99.627278,7314
131,hardship_length,float64,2250055,99.530537,2
124,hardship_reason,object,2250055,99.530537,10
125,hardship_status,object,2250055,99.530537,4
126,deferral_term,float64,2250055,99.530537,2
127,hardship_amount,float64,2250055,99.530537,8951
128,hardship_start_date,object,2250055,99.530537,27


# Conclusion

This notebook completed the initial data understanding step for the Lending Club dataset.

Key observations:
- The dataset is large, with more than 2 million loan records and 145 columns.
- The target column `loan_status` contains multiple repayment outcomes.
- Several columns have very high missing percentages.
- The dataset includes both numerical and categorical features.
- Initial important columns were identified for future feature engineering and modeling.

Next steps:
- Clean and preprocess the dataset
- Remove highly missing or irrelevant columns
- Create a binary default flag
- Engineer borrower risk features
- Prepare the data for machine learning models
